mathics/builtin/list/constructing.py

```diff
@@ -110,7 +111,7 @@ def rec(rest_dims, current):
                     level.append(rec(rest_dims[1:], current + [index]))
                 return Expression(head, *level)
             else:
-                return Expression(f, *(Integer(index) for index in current))
+                return Expression(f, *current, element_conversion_fn=Integer)
```

Comment:
This is the case in which I think it makes sense to tell the constructor that the elements are '_fully_evaluated'  (after the conversion)

Reply:
With the most recent changes and in tracing execution it feels to me that the changes in https://github.com/Mathics3/mathics-core/blob/9c91a7641eaae6536103201dd73f724c55d0c09e/mathics/core/expression.py#L303-L309 pretty much handle detection whether evaluation is needed in the most egregious situations. 

If you can find an example where that is not so, I would be interested in seeing an example.

For me right now, the question is more about how often we will pass a fully formed list and know its properties. Right now I think things are pretty isolated and harmless that it is not worth worrying about too much (unless of course you have an specific example in mind that is likely go occur.)

Reply:
The point is that in this line you create an expression with all the elements "fully evaluated". So, you could pass this information to the constructor, at the time you want to convert the (pythonspeed.com n) ints to (mathics) Integer. In any case, as I told you before, I think this is ready to merge.